src/routegaurded_pages/HomeWithRedux.js

```diff
@@ -0,0 +1,131 @@
+import { useState, useEffect } from "react";
+import { useDispatch, useSelector } from 'react-redux';
+import { getTodos } from "../slices/appSlice";
+import axios from "axios";
+
+export default function HomePage() {
+
+    const [edit, setEdit] = useState({});
+    const dispatch = useDispatch();
+    const todos = useSelector(state => state.app.todos)
+
+    useEffect(() => {
+        dispatch(getTodos());
+    }, []);
+
+    const editTodo = (todo) => {
+        setEdit(todo);
+    }
+
+    const saveTodo = async (todo) => {
+        if(todo === undefined){
+            alert("ENter a value")
+            return
+        }
+        if(todo._id !== undefined){
+            const result = await axios.put(`http://localhost:8000/todos/${todo._id}`,{
+                data: todo,
+                headers: {
+                    'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8'
+                }
+            }).then(resp=>{
+                console.log(resp)
+                alert("Saved")
+            })
+        } else {
+            const result = await axios.post(`http://localhost:8000/todos/`,{
+                data: todo,
+                headers: {
+                    'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8'
+                }
+            }).then(resp=>{
+                console.log(resp)
```

Comment:
Pass a callback function to the redux action. Trigger the callback function once the API call end. Ask me if you don't understand how